### PR TITLE
nixos/prosody: add user/group options, fix pidfile path

### PR DESCRIPTION
**Pull request by brainrake**
###### Motivation for this change

Prosody service options needed improvement
- There were no user/group setting
- The pidfile was in the `dataDir`, not under `/run` as usual

###### Things done

- added `services.prosody.user` and `services.prosody.group` options and implementation
- changed pidfile path, added `RuntimeDirectory` to systemd service config

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by brainrape** [comment]
@GrahamcOfBorg test prosody

**Comment by brainrape** [comment]
@GrahamcOfBorg test prosody

**Comment by brainrape** [comment]
related: #35540 #38643
thanks @gnidorah

**Comment by brainrape** [comment]
@GrahamcOfBorg build prosody

**Review by ghost — APPROVED** [review]
Thanks! Works for me, `nixos/prosody: fix pidfile path` change fixed working when `dataDir` is put under home directory, say `/home/user/prosody`. Before there was an error `Couldn't write pidfile at /home/gnidorah/prosody/prosody.pid; /home/gnidorah/prosody/prosody.pid: Permission denied`

File: nixos/modules/services/networking/prosody.nix [hunk 1]
```diff
@@ -475,15 +487,15 @@ in
         '') cfg.virtualHosts) }
     '';
 
-    users.extraUsers.prosody = {
+    users.extraUsers."${cfg.user}" = {
       uid = config.ids.uids.prosody;
       description = "Prosody user";
       createHome = true;
-      group = "prosody";
+      group = "${cfg.group}";
```
Comment (dotlambda) [inline nixos/modules/services/networking/prosody.nix]: `inherit (cfg) group`

File: nixos/modules/services/networking/prosody.nix [hunk 2]
```diff
@@ -494,9 +506,11 @@ in
       wantedBy = [ "multi-user.target" ];
       restartTriggers = [ config.environment.etc."prosody/prosody.cfg.lua".source ];
       serviceConfig = {
-        User = "prosody";
+        User = "${cfg.user}";
+        Group = "${cfg.user}";
```
Comment (dotlambda) [inline nixos/modules/services/networking/prosody.nix]:
```
User = cfg.user;
Group = cfg.group;
```

File: nixos/modules/services/networking/prosody.nix [hunk 3]
```diff
@@ -475,15 +487,15 @@ in
         '') cfg.virtualHosts) }
     '';
 
-    users.extraUsers.prosody = {
+    users.extraUsers."${cfg.user}" = {
```
Comment (dotlambda) [inline nixos/modules/services/networking/prosody.nix]: I think the user should only be created if `cfg.user == prosody`. Same for the group.

**Review by dotlambda — CHANGES_REQUESTED** [review]

**Comment by brainrape** [comment]
Fixed, thanks for the review!

File: nixos/modules/services/networking/prosody.nix [hunk 4]
```diff
@@ -475,15 +487,15 @@ in
         '') cfg.virtualHosts) }
     '';
 
-    users.extraUsers.prosody = {
+    users.extraUsers.prosody = mkIf (cfg.user == prosody) {
```
Comment (dotlambda) [inline nixos/modules/services/networking/prosody.nix]: prosody should be in quotes

**Comment by brainrape** [comment]
Fixed, tested. Thanks!

**Review by dotlambda — APPROVED** [review]

**Comment by brainrape** [comment]
@GrahamcOfBorg test prosody

**Comment by dotlambda** [comment]
Ran the test manually and can confirm it works.

**Pull request merged**